File: 9663. N Queen.py
# 9663. N-Queen

N = int(input())
chessboard = [0]*N

def check(row):
    for i in range(row):    # 대각선은 거리에 비례
        if chessboard[row] == chessboard[i] or abs(chessboard[row]-chessboard[i]) == abs(row-i):
            return False
    return True

def solve(row):
    global cnt
    if row >= N:
        cnt += 1
    else:
        for column in range(N):
            chessboard[row] = column    # 행 / 열
            if check(row):              # 1차원 배열의 index를 행, value를 열로 배정.
                solve(row+1)
cnt = 0
solve(0)
print(cnt)

File: BOJ/9663. N Queen.js
// # 9663. N-Queen

function findQueens(N) {
	// 이곳에 작성합니다.
    let chessboard = Array(N).fill(0)
    let answer = 0
    
    const check = (row) => {
        for (let i = 0; i < row; i++) {
            if ((chessboard[row] === chessboard[i]) || (Math.abs(chessboard[row]-chessboard[i]) === Math.abs(row-i))) {
                return false
            }
        }
        return true
    }
    
    const solve = (row) => {
        if (row >= N) {
            answer += 1    
        } else {
            for (let column = 0; column < N; column++) {
                chessboard[row] = column
                if (check(row)) {
                    solve(row+1)
                }
            }
        }
    }
    
    solve(0)
	return answer
}

console.log(findQueens(8)) // 92
